complex/ft_verify.c

```diff
@@ -0,0 +1,334 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <float.h>
+
+#include "fabtest.h"
+
+static int fill_index = 0;
+static const char integ_alphabet[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+static const int integ_alphabet_length = (sizeof(integ_alphabet)/sizeof(*integ_alphabet)) - 1;
+
+#define FT_MIN(dst,src)		if ((dst) > (src)) (dst) = (src)
+#define FT_MAX(dst,src)		if ((dst) < (src)) (dst) = (src)
+#define FT_SUM(dst,src)		(dst) += (src)
+#define FT_PROD(dst,src)	(dst) *= (src)
+#define FT_LOR(dst,src)		(dst) = (dst) || (src)
+#define FT_LAND(dst,src)	(dst) = (dst) && (src)
+#define FT_LXOR(dst,src)	(dst) = ((dst) && !(src)) || (!(dst) && (src))
+#define FT_BOR(dst,src)		(dst) |= (src)
+#define FT_BAND(dst,src)	(dst) &= (src)
+#define FT_BXOR(dst,src)	(dst) ^= (src)
+#define FT_COPY(dst,src)	(dst) = (src)
+
+#define FT_CSWAP(dst,src,cmp,res,cnt,CMP_OP,TYPE)	\
+	do {						\
+		int i;					\
+		TYPE *d = (dst);			\
+		TYPE *s = (src);			\
+		TYPE *c = (cmp);			\
+		TYPE *r = (res);			\
+		for (i = 0; i < (cnt); i++) {		\
+			r[i] = d[i];			\
+			if (c[i] CMP_OP d[i])		\
+				d[i] = s[i];		\
+		}					\
+	} while (0)
+
+#define FT_MSWAP(dst,src,cmp,res,cnt,TYPE)			\
+	do {							\
+		int i;						\
+		TYPE *d = (dst);				\
+		TYPE *s = (src);				\
+		TYPE *c = (cmp);				\
+		TYPE *r = (res);				\
+		for (i = 0; i < (cnt); i++) {			\
+			r[i] = d[i];				\
+			d[i] = (s[i] & c[i]) | (d[i] & ~c[i]);	\
+		}						\
+	} while (0)
+
+#define FT_FILL(dst,src,alph_index,cnt,TYPE)				\
+	do {								\
+		int i;							\
+		TYPE *d = (dst);					\
+		for (i = 0; i < cnt; i++) {				\
+			d[i] = (TYPE) (integ_alphabet[alph_index]);	\
+			if (alph_index >= integ_alphabet_length)	\
+				alph_index = 0;				\
+		}							\
+	} while (0)
+
+#define CASE_INT_TYPE(FUNC,...)					\
+	case FI_INT8:	FUNC(__VA_ARGS__,int8_t); break;	\
+	case FI_UINT8:	FUNC(__VA_ARGS__,uint8_t); break;	\
+	case FI_INT16:	FUNC(__VA_ARGS__,int16_t); break;	\
+	case FI_UINT16: FUNC(__VA_ARGS__,uint16_t); break;	\
+	case FI_INT32:	FUNC(__VA_ARGS__,int32_t); break;	\
+	case FI_UINT32: FUNC(__VA_ARGS__,uint32_t); break;	\
+	case FI_INT64:	FUNC(__VA_ARGS__,int64_t); break;	\
+	case FI_UINT64: FUNC(__VA_ARGS__,uint64_t); break;
+
+#define CASE_FP_TYPE(FUNC,...)					\
+	case FI_FLOAT:	FUNC(__VA_ARGS__,float); break;		\
+	case FI_DOUBLE:	FUNC(__VA_ARGS__,double); break;	\
+	case FI_LONG_DOUBLE: FUNC(__VA_ARGS__,long double); break;
+
+#define CASE_COMPLEX_TYPE(FUNC,...) \
+	case FI_FLOAT_COMPLEX:	FUNC(__VA_ARGS__,float complex); break;		\
+	case FI_DOUBLE_COMPLEX:	FUNC(__VA_ARGS__,double complex); break;	\
+	case FI_LONG_DOUBLE_COMPLEX: FUNC(__VA_ARGS__,long double complex); break;
```

Comment:
Please pull in the atomics and complex datatypes and operations from libfabric for portability.